### DISCUSSION TITLE
How do I mock the specific class method from external module?

### DISCUSSION THREAD

**Discussion by Medvedscak** [webdriverio/webdriverio#11819] [Q&A] [answered]
I've read this page in docs but it didn't help me: https://webdriver.io/docs/component-testing/mocking/

In my case I have a module with several exports, one of which is a class. In my component tests in order to make localizations work, I need to mock one method of this class. Here's an example:

```
// userPreferencesModule.js

export declare class userPreferencesInstance {
  static getUserLocale(): Promise<string> {
    // implementation here
  }

  // Other methods...
}

// Exporting other constants or functionalities
export const CONSTANT_ONE = 'CONSTANT_ONE';
export const CONSTANT_TWO = 'CONSTANT_TWO';
```

I'd like to use the mocked version of this userPreferencesInstance.getUserLocale().
And another thing is that it's not used directly in tests but it's the setup which should be done for the component test before running those tests. I'm not sure what's the best place to call this - before hook in wdio.conf or in the suite itself or somewhere else (like mochaOpts require).

Please advise,
thanks

**Comment by christian-bromann** [comment]
> And another thing is that it's not used directly in tests but it's the setup which should be done for the component test before running those tests.

Can you elaborate?

**Reply by Medvedscak** [reply]
The initial issue is that the react component uses localizations which are not working when I render the component in tests. The dev team is using jest for unit testing and they use jest moduleNameMapper (https://jestjs.io/docs/configuration#modulenamemapper-objectstring-string--arraystring) to mock the module during the jest setup.
I'm trying do do something similar with webdriverio. This will probably work in both webdriverio before hook and in mocha before hook but I'm current;y trying to understand how to mock the specific class method only.

**Reply by christian-bromann** [reply, marked as answer]
Gotcha, you can do this in the `wdio.config.js` using Vite's [alias](https://vitejs.dev/config/shared-options.html#resolve-alias) option, e.g.

```ts
export const config: WebdriverIO.Config = {
    runner: [
        'browser',
        {
            preset: 'react',
            viteConfig" {
                alias: {
                    'foo': 'bar'
                }
            }
        }
    ],
    // ...
}
```

With this example importing `foo` will resolve to `bar`.

**Reply by Medvedscak** [reply]
Thanks, it solved the half of the problems) I was even using this vite config alias for resolving import aliases, but didn't think of using it for this purpose)

**Reply by christian-bromann** [reply]
Which other half is unsolved?

**Reply by Medvedscak** [reply]
this is probably something related to internal project logic. I see that mocks are working fine (the locale is correct) but for some reason the component is rendered with localization keys instead of localized strings. Need to debug more but in unit tests with jest it works with just mocking some methods (I'm doing absolutely the same). Can't say for now if it's something else should be mocked or maybe some import is not resolved correctly or missing (no errors though, just missing localizations)

**Reply by Medvedscak** [reply]
@christian-bromann aliases work fine when I was mocking real aliases (@app/something) and packages. But now I'm checking a simple case for debugging and I'm trying just to mock one of own modules with the same but mocked.
I.e. I have this in **src/test-setup.ts**:
`import { UserPreferences } from "./user-preferences";`

And in vite.config.ts I'm trying to mock this **src/user-preferences** with **src/__mock__/user-preferences** like this:
```
...
alias: {
      "./user-preferences": path.resolve(
        __dirname,
        "./__mock__/user-preferences"
      ),
    },
...

```
I've read that I should use the exact relative import string the same way it's used in the module where I'm going to mock it. That's why I'm using **"./user-preferences"**. I was trying different options for the path of the mocked module for vite config but nothing works. Maybe you know how to use it correctly?
